src/service/async_service_impl.cc

```diff
@@ -0,0 +1,168 @@
+  // Used to send the GRPC response
+  grpc::ServerAsyncResponseWriter<CheckResponse> responder_;
+
+  // Boost::ASIO I/O service
+  boost::asio::io_context &io_context_;
+
+  std::shared_ptr<authservice::config::Config> config_;
+  authservice::service::AuthServiceImpl impl_;
+};
+
+void CompleteState::Proceed() {
+  spdlog::trace("Processing completion and deleting state");
+
+  delete processor_;
+  delete this;
+}
+
+AsyncAuthServiceImpl::AsyncAuthServiceImpl(std::shared_ptr<authservice::config::Config> config)
+        : config_(std::move(config)),
+          io_context_(std::make_shared<boost::asio::io_context>()) {
+  grpc::ServerBuilder builder;
+  builder.AddListeningPort(config::GetConfiguredAddress(*config_), grpc::InsecureServerCredentials());
+  builder.RegisterService(&service_);
+  cq_ = builder.AddCompletionQueue();
+  server_ = builder.BuildAndStart();
+}
+
+void AsyncAuthServiceImpl::Run() {
+  // Add a work object to the IO service so it will not shut down when it has nothing left to do
+  auto work = std::make_shared<boost::asio::io_context::work>(*io_context_);
+
+  // Spin up our worker threads
+  // Config validation should have already ensured that the number of threads is > 0
+  boost::thread_group threadpool;
+  for (unsigned int i = 0; i < config_->threads(); ++i) {
+    threadpool.create_thread([this](){
+      while(true) {
+        try {
+          this->io_context_->run();
+          break;
+        } catch(std::exception & e) {
+          spdlog::error("Unexpected error in worker thread: {}", e.what());
+        }
+      }
+    });
+  }
+
+  spdlog::info("{}: Server listening on {}", __func__, config::GetConfiguredAddress(*config_));
+
+  try {
+    // Spawn a new state instance to serve new clients.
+    new ProcessingState(config_, service_, *cq_, *io_context_);
+
+    void *tag;
+    bool ok;
+    while (true) {
+      // Block waiting to read the next event from the completion queue. The
+      // event is uniquely identified by its tag, which in this case is the
+      // memory address of a CallData instance.
+      // The return value of Next should always be checked. This return value
+      // tells us whether there is any kind of event or cq_ is shutting down.
+      if (!cq_->Next(&tag, &ok)) {
+        break;
+      }
+      GPR_ASSERT(ok);
+
+      static_cast<ServiceState *>(tag)->Proceed();
+    }
+  } catch (const std::exception &e) {
+    spdlog::error("{}: Unexpected error: {}", __func__, e.what());
+  }
+
+  spdlog::info("Server shutting down");
+
+  // Start shutting down gRPC
+  server_->Shutdown();
+  cq_->Shutdown();
+
+  // Continue calling Next to process any outstanding events.
+  try {
+    void *tag;
+    bool ok;
+    while (true) {
+      if (!cq_->Next(&tag, &ok)) {
+        break;
+      }
+      GPR_ASSERT(ok);
+
+      static_cast<ServiceState *>(tag)->Proceed();
+    }
+  } catch (const std::exception &e) {
+    spdlog::error("{}: Unexpected error: {}", __func__, e.what());
+  }
```

Comment:
Perhaps this could have a little more comments?  We're curious what conditions this is intended to handle.

It seems like you could arrive here if the previous loop caught any kind of exception or `Next` returned false.
  - If `Next` returned false in the first loop, would we expect it too also return false in the second loop?
  - On the other hand, if we arrive in the second loop because of an exception, should all kinds of exceptions lead to outstanding events being processed?

Reply:
This is mainly to work with a peculiarity of the gRPC completion queue. The second loop is just to drain the queue after shutting it down, because if the completion queue's destructor is called before draining any outstanding events it triggers an assertion. I'll add some additional comments to explain this.